wallet/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+pub mod error;
+
+use std::collections::{BTreeMap, HashSet};
+
+pub use error::WalletError;
+use nomos_core::{
+    block::Block,
+    header::HeaderId,
+    mantle::{keys::PublicKey, ledger::Tx as LedgerTx, AuthenticatedMantleTx, NoteId, Utxo, Value},
+};
+use nomos_ledger::LedgerState;
+
+pub struct WalletBlock {
+    pub id: HeaderId,
+    pub parent: HeaderId,
+    pub ledger_txs: Vec<LedgerTx>,
+}
+
+impl<Tx: AuthenticatedMantleTx> From<Block<Tx>> for WalletBlock {
+    fn from(block: Block<Tx>) -> Self {
+        Self {
+            id: block.header().id(),
+            parent: block.header().parent(),
+            ledger_txs: block
+                .transactions()
+                .map(|auth_tx| auth_tx.mantle_tx().ledger_tx.clone())
+                .collect(),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct WalletState {
+    utxos: rpds::HashTrieMapSync<NoteId, Utxo>,
+    pk_index: rpds::HashTrieMapSync<PublicKey, rpds::HashTrieSetSync<NoteId>>,
+}
+
+impl WalletState {
+    pub fn from_ledger(known_keys: &HashSet<PublicKey>, ledger: &LedgerState) -> Self {
+        let mut utxos = rpds::HashTrieMapSync::new_sync();
+        let mut pk_index = rpds::HashTrieMapSync::new_sync();
+
+        for (_, (utxo, _)) in ledger.latest_commitments().utxos().iter() {
+            if known_keys.contains(&utxo.note.pk) {
+                let note_id = utxo.id();
+                utxos = utxos.insert(note_id, *utxo);
+
+                let note_set = pk_index
+                    .get(&utxo.note.pk)
+                    .cloned()
+                    .unwrap_or_else(rpds::HashTrieSetSync::new_sync)
+                    .insert(note_id);
+                pk_index = pk_index.insert(utxo.note.pk, note_set);
+            }
+        }
+
+        Self { utxos, pk_index }
+    }
+
+    pub fn utxos_for_amount(
+        &self,
+        amount: Value,
+        pks: impl IntoIterator<Item = PublicKey>,
```

Comment:
is this collecting only utxos with these pks? What's the intended usage for a wallet (like, aren't all utxos in here already under user control) ?

Reply:
> is this collecting only utxos with these pks? 

Yep

> What's the intended usage for a wallet

I want eventually provide a CLI that will connect to this wallet service. Then all the normal account management ideas apply. e.g. pk for "business", pk for "personal" or whatever accounting system you want to use.

Reply:
could you instantiate different wallet instances instead?